jobs/fix_crt_channel_unreads/worker.go

```diff
@@ -0,0 +1,260 @@
+	App *app.App
+}
+
+func init() {
+	app.RegisterFixCRTChannelUnreadsJobInterface(func(s *app.Server) tjobs.FixCRTChannelUnreadsJobInterface {
+		a := app.New(app.ServerConnector(s))
+		return &FixCRTChannelUnreadsJobInterfaceImpl{a}
+	})
+}
+
+func (i *FixCRTChannelUnreadsJobInterfaceImpl) MakeWorker() model.Worker {
+
+	return &FixCRTChannelUnreadsWorker{
+		name:        JobName,
+		stopChan:    make(chan struct{}),
+		stoppedChan: make(chan struct{}),
+		jobsChan:    make(chan model.Job),
+		jobServer:   i.App.Srv().Jobs,
+		app:         i.App,
+	}
+}
+
+func (w *FixCRTChannelUnreadsWorker) JobChannel() chan<- model.Job {
+	return w.jobsChan
+}
+
+func (w *FixCRTChannelUnreadsWorker) Run() {
+	mlog.Debug("Worker started", mlog.String("worker", w.name))
+
+	// kill all in-progress jobs in DB. This can happen if server
+	// was shut down incorrectly
+	olderJobs, err := w.app.Srv().Store.Job().GetAllByStatus(model.JobStatusInProgress)
+	if err == nil {
+		for _, j := range olderJobs {
+			if j.Type == model.JobTypeFixChannelUnreadsForCRT {
+				w.setJobError(
+					j,
+					model.NewAppError(w.name, "fix_crt_channel_unreads.worker.do_job.kill_orphan_jobs",
+						nil,
+						"killing orphan jobs",
+						http.StatusInternalServerError),
+				)
+			}
+		}
+	}
+
+	defer func() {
+		mlog.Debug("Worker finished", mlog.String("worker", w.name))
+		close(w.stoppedChan)
+	}()
+
+	for {
+		select {
+		case <-w.stopChan:
+			mlog.Debug("Worker received stop signal", mlog.String("worker", w.name))
+			return
+		case job := <-w.jobsChan:
+			mlog.Debug("Worker received a new candidate job.", mlog.String("worker", w.name))
+			w.doJob(&job)
+		}
+	}
+}
+
+func (w *FixCRTChannelUnreadsWorker) Stop() {
+	mlog.Debug("Worker stopping", mlog.String("worker", w.name))
+	close(w.stopChan)
+	<-w.stoppedChan
+}
+
+func (w *FixCRTChannelUnreadsWorker) doJob(job *model.Job) {
+	if claimed, err := w.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("Worker experienced an error while trying to claim job",
+			mlog.String("worker", w.name),
+			mlog.String("job_id", job.Id),
+			mlog.String("error", err.Error()))
+		return
+	} else if !claimed {
+		return
+	}
+	if _, err := w.app.Srv().Store.System().GetByName(model.MigrationKeyFixCRTChannelUnreads); err == nil {
+		mlog.Info("Worker: migration already done", mlog.String("worker", w.name), mlog.String("job_id", job.Id))
+		w.setJobSuccess(job)
+		return
+	}
+
+	fixedBadCM := 0
+	migrationDone := false
+	prevErr := false
+	userID, channelID := w.getProgressFromPreviousJobs(job)
+	if userID != "" && channelID != "" {
+		mlog.Info("Restarting from previous job run", mlog.String("UserID", userID), mlog.String("ChannelID", channelID))
+	}
+	shouldStop := false
+	for {
+		select {
+		case <-w.stopChan:
+			shouldStop = true
+		default:
+			shouldStop = false
+		}
+		if shouldStop {
+			break
+		}
+		cms, sErr := w.app.Srv().Store.Channel().GetCRTUnfixedChannelMembershipsAfter(userID, channelID, 100)
```

Comment:
You mentioned that on your test environment this job took around 3 minutes. Were you able to monitor the impact of this on DB CPU? I am just wondering if we should artificially limit the throughput to avoid stressing the server too much given this is meant to be run on live systems.

Reply:
>you mentioned that on your test environment this job took around 3 minutes
The previous column addition takes three minutes.  This whole job takes a very long time (scales with the number of bad channel memberships). 

>Were you able to monitor the impact of this on DB CPU?

There is indeed a noticeable spike in the DB CPU usage. I haven't noticed any slow-down in my test server usage, but I don't have any solid numbers to measure user-level impact.


>I am just wondering if we should artificially limit the throughput to avoid stressing the server too much given this is meant to be run on live systems.

Yep, I was thinking about this as well. What would be the best way to achieve this? Will a simple sleep to do it? I wasn't sure how we could get some sort of back pressure from the connection pool or DB.

Reply:
> There is indeed a noticeable spike in the DB CPU usage. I haven't noticed any slow-down in my test server as a consequence, but I don't have any solid numbers to measure user-level impact.

Okay, @isacikgoz and I can probably assist on this front. It would be good to setup a new test cluster with some good amount of data (e.g. 10M posts) and fake some bad memberships. As a real datapoint, we have around 100k in community db right now.

> Yep, I was thinking about this as well. What would be the best way to achieve this? Will a simple sleep to do it? I wasn't sure how we could get some sort of back pressure from the connection pool or DB.

I think simple sleeping could work but will need some testing to tune it properly. To improve things further I think we should also consider a couple other things:

- Amount of returned memberships. It's currently set to 100 but increasing this a bit could help minimizing read operations. Need to be tested of course.
- Batching write operations. `ChannelStore.UpdateLastViewedAt()` is one of the most complex store calls. We should first consider if we actually need it as it does a lot more than what it says. If the answer is yes, meaning we need all the logic attached to it then I'd suggest trying to call it the least possible number of times by passing multiple channel ids at once.